### ISSUE TITLE
PermissionDenied while accessing secrets

### ISSUE THREAD

**Issue by wojcikm** [GoogleCloudPlatform/berglas#150]
Hi, I have done setup as the documentation specifies [here](https://github.com/GoogleCloudPlatform/berglas) and [here](https://github.com/GoogleCloudPlatform/berglas/tree/main/examples/kubernetes). Everything sets up, but when container with injected values from Secret Manager starts, I get the following error:

```text
failed to access secret sm://project-id/SECRET_KEY: failed to access secret: rpc error: code = PermissionDenied 
desc = Permission 'secretmanager.versions.access' denied for resource 
'projects/project-id/secrets/SECRET_KEY/versions/latest' (or it may not exist).
```

I have double checked the service accounts and role bindings and cannot find any clues. SA has `Secret Manager Admin` role  for testing purposes.

Maybe there is issue with this `/versions/latest` suffix? When I try to get it with it via `berglas access` cli I get:

```text
$ berglas access sm://project-id/SECRET_KEY/version/latest
failed to parse reference "sm://project-id/SECRET_KEY/version/latest": 
invalid secret name "SECRET_KEY/version/latest"
```

But when using it without, it just works fine:
```text
$ berglas access sm://project-id/SECRET_KEY
secret_value
```

**Comment by sethvargo** [comment]
It's `versions` (plural). Does that work for you?

**Comment by wojcikm** [comment]
Yeah, you are right. It should be plural - however error persists.

**Comment by sethvargo** [comment]
For accessing the version, it should be using the # key

**Comment by wojcikm** [comment]
I have changed retrieving a key to `sm://project-id/SECRET_KEY#1`, but still have the same error:

```
failed to access secret sm://project-id/SECRET_KEY#1: failed to access secret: rpc error: code = PermissionDenied desc = Permission 'secretmanager.versions.access' denied for resource 'projects/project-id/secrets/SECRET_KEY/versions/1' (or it may not exist).
```

But via cli it works:
```
$ berglas access sm://project-id/SECRET_KEY#1
secret_value
```

**Comment by sethvargo** [comment]
Can you tell me more about how you're running your workload? What is the underlying service account? What permissions does it have?

**Comment by wojcikm** [comment]
Service account in GCP:
```
berglas-accessor@bks-dev-1-eu-w1.iam.gserviceaccount.com Secret Manager Admin
```

Service account in k8s:
```
envserver
```

Policy binding:
```
gcloud iam service-accounts add-iam-policy-binding   
--project ${PROJECT_ID}   
--role "roles/iam.workloadIdentityUser"   
--member "serviceAccount:${PROJECT_ID}.svc.id.goog[default/envserver]" berglas-accessor@${PROJECT_ID}.iam.gserviceaccount.com
```

Deployment env:
```
- name: SECRET_KEY
     value: sm://<PROJECT_ID>/SECRET_KEY#1
```

**Comment by sethvargo** [comment]
Did you deploy the mutating webhook controller? What does the pod look like after you deploy (`kubectl describe po envserver -o yaml`)? Did you enable Workload Identity on the GKE cluster when you created it? Did you annotate the K8S service account with the name of the Google Cloud service account?

**Comment by sethvargo** [comment]
Hmmmm - can you show me the full logs from the container?

**Comment by wojcikm** [comment]
This is the error I see in logs, nothing more:

> 
> ```
> failed to access secret sm://project-id/SECRET_KEY: failed to access secret: rpc error: code = PermissionDenied 
> desc = Permission 'secretmanager.versions.access' denied for resource 
> 'projects/project-id/secrets/SECRET_KEY/versions/latest' (or it may not exist).
> ```

**Comment by sethvargo** [comment]
Does the error message literally say "project-id" or are you replacing the values?

**Comment by wojcikm** [comment]
I am replacing all sensitive values, obviously :)

**Comment by sethvargo** [comment]
Do you know if you have VPC-SC, other security perimeters, or org policies that would restrict access? Sorry, it's basically impossible to debug these things 😦 

Can you try running a cloudsdk container on your cluster as that k8s SA and exec into it, and run `gcloud secrets access`?

**Comment by wojcikm** [comment]
There are no perimeters in `VPC Service Controls`.

Container with cloud sdk:
```
root@cloud-sdk-cdd954df6-zjtrf:/# gcloud secrets list
NAME                 CREATED                         REPLICATION_POLICY  LOCATIONS
SECRET_KEY    2021-04-20T21:43:55  automatic           -

```

**Comment by sethvargo** [comment]
Hmm so if it's working with gcloud, there's no reason it shouldn't be working inside the container.

I don't see `serviceAccountName` in the YAML spec - are you sure it's set on the workload?

**Comment by wojcikm** [comment]
Yeah, it is there. But what is more interesting, even if I delete `serviceAccountName` I am still able to get secrets via `gcloud secrets list`

**Comment by sethvargo** [comment]
Is it in the same namespace?

**Comment by wojcikm** [comment]
Yes it is.

**Comment by wojcikm** [comment]
Yes it is.

**Comment by wojcikm** [comment]
It looks like k8s does not use binded SA:

```
$ kubectl run -it \
>   --image google/cloud-sdk:slim \
>   --serviceaccount envserver \
>   --namespace default \
>   workload-identity-test

If you don't see a command prompt, try pressing enter.

root@workload-identity-test:/# gcloud auth list
                  Credentialed Accounts
ACTIVE  ACCOUNT
*       111111111-compute@developer.gserviceaccount.com

To set the active account, run:
    $ gcloud config set account `ACCOUNT
```

Workload identity is enabled on cluster, bindings are double-checked. Permissions are granted via `berglas` cli.

**Comment by wojcikm** [comment]
Finally we got to the point. Hence I was doing poc on existing cluster and even though I enabled worklad identity on it, I have missed [that nodes have to be updated ](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#migrate_applications_to) to utilize it. Now it works like a charm! 

So the punchline is: update existing nodes to use workload identity:
```
gcloud container node-pools update NODEPOOL_NAME \
  --cluster=CLUSTER_NAME \
  --workload-metadata=GKE_METADATA
```
Thanks for your support @sethvargo  and great tool!
